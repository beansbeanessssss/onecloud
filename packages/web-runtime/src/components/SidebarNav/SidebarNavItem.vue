<template>
  <li
    ref="item"
    class="oc-sidebar-nav-item oc-pb-xs oc-px-s"
    :aria-current="active ? 'page' : null"
  >
    <oc-button
      v-oc-tooltip="toolTip"
      type="router-link"
      :appearance="active ? 'raw-inverse' : 'raw'"
      :class="['oc-sidebar-nav-item-link', { active: active }]"
      :to="target"
      :data-nav-id="index"
      :data-nav-name="navName"
    >
      <span class="oc-flex">
        <oc-icon :name="icon" :fill-type="fillType" />
        <span class="oc-ml-m text" :class="{ 'text-invisible': collapsed }" v-text="name" />
      </span>
      <oc-tag v-if="tag" class="oc-py-rm" size="small">{{ tag }}</oc-tag>
    </oc-button>
  </li>
</template>
<script lang="ts">
import { defineComponent } from 'vue'
import { Router } from 'vue-router'

export default defineComponent({
  props: {
    name: {
      type: String,
      required: true
    },
    index: {
      type: String,
      required: true
    },
    active: {
      type: Boolean,
      required: false,
      default: false
    },
    target: {
      type: String,
      required: false,
      default: null
    },
    icon: {
      type: String,
      required: true
    },
    fillType: {
      type: String,
      required: false,
      default: 'fill'
    },
    collapsed: {
      type: Boolean,
      required: false,
      default: false
    },
    tag: {
      type: String,
      required: false,
      default: null
    }
  },
  computed: {
    navName() {
      return (this.$router as Router)?.resolve(this.target, this.$route)?.name || 'route.name'
    },
    toolTip() {
      const value = this.collapsed
        ? this.$gettextInterpolate(this.$gettext('Navigate to %{ pageName } page'), {
            pageName: this.name
          })
        : ''
      return {
        content: value,
        placement: 'right',
        arrow: false
      }
    }
  },
  watch: {
    active(active) {
      if (!active) {
        return
      }
      this.animateHighlightPosition(this.index)
    }
  },
  mounted() {
    if (!this.active) {
      return
    }
    this.animateHighlightPosition(this.index)
  },
  methods: {
    animateHighlightPosition(target, durationSeconds = 0.2) {
      const highlightedElement = document.getElementById('nav-highlighter')
      if (!highlightedElement) {
        return
      }
      const targetElement = this.$refs.item
      const offset = targetElement.offsetTop
      const style = highlightedElement.style
      style.setProperty('transition-duration', `${durationSeconds}s`)
      style.setProperty('transform', `translateY(${offset}px)`)
    }
  }
})
</script>

<style lang="scss">
.oc-sidebar-nav-item-link {
  position: relative;
  align-items: center !important;
  display: flex !important;
  justify-content: space-between !important;
  padding: var(--oc-space-small) !important;
  border-radius: 5px;
  white-space: nowrap;
  user-select: none;

  .oc-tag {
    color: var(--oc-color-text-default);
    background-color: var(--oc-color-background-highlight);
  }
  .text {
    opacity: 1;
    transition: all 0.3s;
  }
  .text-invisible {
    opacity: 0 !important;
    transition: 0s;
  }

  &:hover:not(.active) {
    color: var(--oc-color-swatch-brand-hover) !important;
  }

  &:hover,
  &:focus {
    text-decoration: none !important;
  }
  &.active {
    overflow: hidden;
  }

  .oc-icon svg {
    transition: all 0.3s;
  }
}
</style>
