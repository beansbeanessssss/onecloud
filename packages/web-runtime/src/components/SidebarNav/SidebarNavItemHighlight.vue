<template>
  <span v-if="initial" id="highlighter" :data-initial-id="index" />
</template>
<script>
export default {
  props: {
    index: {
      type: Number,
      required: true
    },
    active: {
      type: Boolean,
      required: true
    }
  },
  data() {
    return {
      initial: false
    }
  },
  watch: {
    active(active) {
      if (!active) return
      this.animateHighlightPosition(this.index)
    }
  },
  created() {
    if (!this.active) return
    this.initial = true
  },
  methods: {
    animateHighlightPosition(target, durationSeconds = 0.2) {
      const initialId = this.getInitialId()
      const currentElement = this.getNavigationItem(initialId)
      const targetElement = this.getNavigationItem(target)
      const distanceAbs = this.getDistanceBetweenElements(currentElement, targetElement)
      const distance = target < initialId ? -distanceAbs : distanceAbs
      const style = this.getHighlightingElement().style
      style.setProperty('transition-duration', `${durationSeconds}s`)
      style.setProperty('transform', `translateY(${distance}px)`)
    },
    getNavigationItem(index) {
      return document.querySelectorAll(`[data-nav-id="${index}"]`)[0]
    },
    getPositionAtCenter(element) {
      const { top, left, width, height } = element.getBoundingClientRect()
      return {
        x: left + width / 2,
        y: top + height / 2
      }
    },
    getDistanceBetweenElements(a, b) {
      const aPosition = this.getPositionAtCenter(a)
      const bPosition = this.getPositionAtCenter(b)
      return Math.hypot(aPosition.x - bPosition.x, aPosition.y - bPosition.y)
    },
    getHighlightingElement() {
      return document.getElementById('highlighter')
    },
    getInitialId() {
      return this.getHighlightingElement().getAttribute('data-initial-id')
    }
  }
}
</script>

<style lang="scss">
#highlighter {
  position: absolute;
  top: 0;
  left: 0;
  right: 0;
  bottom: 0;
  background: linear-gradient(90deg, #0869de 0%, #4e85c8 100%);
  transition: transform 0.2s cubic-bezier(0.51, 0.06, 0.56, 1.37);
  border-radius: 5px;
  z-index: -1;
}
#highlighter::before {
	content: "";
	position: absolute;
	width: 45px;
	top: 0;
	bottom: 0;
	left: 0;
	box-shadow: 2px 0px 6px rgba(0, 0, 0, 0.14);
}
</style>
