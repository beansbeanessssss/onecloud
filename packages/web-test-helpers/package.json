{
  "name": "web-test-helpers",
  "private": true,
  "main": "src/index.ts",
  "peerDependencies": {
    "axios": "0.27.2",
    "vue-gettext": "2.1.12",
    "vue-router": "3.6.5",
    "vuex": "4.1.0"
  }
}
