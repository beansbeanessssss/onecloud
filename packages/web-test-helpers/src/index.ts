export * from './mocks/defaultComponentMocks'
export * from './mocks/store/defaultStoreMockOptions'
export * from './mocks/defaultStubs'
export * from './mocks/axios'
export * from './defaultPlugins'
export * from './helpers'
