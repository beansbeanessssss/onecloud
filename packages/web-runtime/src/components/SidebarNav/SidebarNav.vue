<template>
  <div
    id="web-nav-sidebar"
    :class="['uk-visible@l', { 'oc-app-navigation-collapsed': navigation.closed }]"
  >
    <oc-button
      variation="inverse"
      appearance="raw"
      :class="toggleSidebarButtonClass"
      class="toggle-sidebar-button oc-py-s"
      :aria-label="$gettext('Toggle sidebar')"
      @click="toggleSidebarButtonClick"
    >
      <oc-icon size="large" fill-type="line" :name="toggleSidebarButtonIcon" />
    </oc-button>
    <nav
      class="oc-sidebar-nav oc-my-l oc-px-xs"
      :aria-label="$gettext('Sidebar top navigation menu')"
    >
      <oc-list>
        <sidebar-nav-item
          v-for="link in navItems"
          :key="link.route.path"
          :target="link.route.path"
          :active="link.active"
          :icon="link.icon || link.iconMaterial"
          :fill-type="link.fillType"
          :name="link.name"
          :collapsed="navigation.closed"
        />
      </oc-list>
    </nav>
    <!-- @slot bottom content of the sidebar -->
    <slot name="bottom" />
  </div>
</template>

<script>
import { mapState, mapActions } from 'vuex'
import SidebarNavItem from './SidebarNavItem.vue'

export default {
  components: {
    SidebarNavItem
  },
  props: {
    navItems: {
      type: Array,
      required: true
    }
  },
  computed: {
    ...mapState(['sidebar', 'navigation']),

    toggleSidebarButtonClass() {
      return this.navigation.closed
        ? 'toggle-sidebar-button-collapsed'
        : 'toggle-sidebar-button-expanded oc-pr-s'
    },

    toggleSidebarButtonIcon() {
      return this.navigation.closed ? 'arrow-drop-right' : 'arrow-drop-left'
    }
  },
  methods: {
    ...mapActions(['openNavigation', 'closeNavigation']),

    toggleSidebarButtonClick() {
      return this.navigation.closed ? this.openNavigation() : this.closeNavigation()
    }
  }
}
</script>

<style lang="scss">
#web-nav-sidebar {
  background-color: #2d2d2d;
  border-top-left-radius: 15px;
  border-bottom-left-radius: 15px;
  color: var(--oc-color-text-inverse);
  display: flex;
  flex-direction: column;
  position: sticky;
  top: 0;
  transition: all 0.35s cubic-bezier(0.34, 0.11, 0, 1.12);
  width: var(--oc-size-width-medium);

  .toggle-sidebar-button {
    transition: all 0.2s ease-out;
    &:hover {
      border-top-left-radius: 15px;
      background: #383838;
    }
  }
  .toggle-sidebar-button-expanded {
    justify-content: flex-end !important;
  }
}

.oc-app-navigation-collapsed {
  width: 75px !important;
}
</style>
