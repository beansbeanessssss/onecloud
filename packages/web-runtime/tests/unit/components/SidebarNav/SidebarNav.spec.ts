import SidebarNav from 'web-runtime/src/components/SidebarNav/SidebarNav.vue'
import sidebarNavItemFixtures from '../../../__fixtures__/sidebarNavItems'
import { createStore, defaultPlugins, defaultStubs, mount } from 'web-test-helpers'

jest.mock('uuid', () => ({
  v4: () => {
    return '00000000-0000-0000-0000-000000000000'
  }
}))

const slots = {
  bottom: '<span class="footer">Footer</span>'
}

describe('OcSidebarNav', () => {
  const { wrapper } = getWrapper()

  it('displays a bottom slot if given', () => {
    expect(wrapper.findAll('.footer').length).toBe(1)
  })

  it('renders navItems into a list', () => {
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('toggles into closed state upon button click', async () => {
    await wrapper.find('.toggle-sidebar-button').trigger('click')
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })

  it('toggles back into open state upon button click', async () => {
    await wrapper.find('.toggle-sidebar-button').trigger('click')
    await wrapper.vm.$nextTick()
    expect(wrapper.html()).toMatchSnapshot()
  })
})

function getWrapper() {
  const storeOptions = {
    state: {
      navigation: {
        closed: false
      }
    },
    mutations: {
      SET_CLOSED(state, value) {
        state.navigation.closed = value
      }
    },
    actions: {
      openNavigation({ commit }) {
        commit('SET_CLOSED', false)
      },
      closeNavigation({ commit }) {
        commit('SET_CLOSED', true)
      }
    }
  }
  const store = createStore(storeOptions)
  return {
    wrapper: mount(SidebarNav, {
      slots,
      props: {
        navItems: sidebarNavItemFixtures
      },
      global: {
        renderStubDefaultSlot: true,
        plugins: [...defaultPlugins(), store],
        stubs: {
          ...defaultStubs,
          'sidebar-nav-item': true,
          OcButton: false,
          'oc-icon': true,
          'oc-list': true
        }
      }
    })
  }
}
