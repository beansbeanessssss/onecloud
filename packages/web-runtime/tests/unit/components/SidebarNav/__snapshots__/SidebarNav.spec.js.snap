// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OcSidebarNav renders navItems into a list 1`] = `
<div id="web-nav-sidebar" class="oc-app-navigation-expanded"><button aria-label="Toggle sidebar" class="toggle-sidebar-button oc-py-s oc-button oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-inverse oc-button-inverse-raw toggle-sidebar-button-expanded oc-pr-s">
    <oc-icon-stub name="arrow-drop-left" filltype="line" accessiblelabel="" type="span" size="large" variation="passive"></oc-icon-stub>
  </button>
  <nav aria-label="Sidebar top navigation menu" class="oc-sidebar-nav oc-my-l oc-px-xs">
    <oc-list-stub>
      <sidebar-nav-item-stub name="All files" index="0" active="true" target="/files/list/all" icon="folder" filltype="fill"></sidebar-nav-item-stub>
      <sidebar-nav-item-stub name="Shared with me" index="1" target="/files/list/shared-with-me" icon="shared-with-me" filltype="fill"></sidebar-nav-item-stub>
      <sidebar-nav-item-stub name="Shared with others" index="2" target="/files/list/shared-with-others" icon="shared-with-others" filltype="fill"></sidebar-nav-item-stub>
      <sidebar-nav-item-stub name="Shared via link" index="3" target="/files/list/shared-via-link" icon="link" filltype="fill"></sidebar-nav-item-stub>
      <sidebar-nav-item-stub name="Deleted files" index="4" target="/files/list/trash-bin" icon="delete" filltype="fill"></sidebar-nav-item-stub>
    </oc-list-stub>
  </nav> <span class="footer">Footer</span>
</div>
`;

exports[`OcSidebarNav toggles back into open state upon button click 1`] = `
<div id="web-nav-sidebar" class="oc-app-navigation-expanded"><button aria-label="Toggle sidebar" class="toggle-sidebar-button oc-py-s oc-button oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-inverse oc-button-inverse-raw toggle-sidebar-button-expanded oc-pr-s">
    <oc-icon-stub name="arrow-drop-left" filltype="line" accessiblelabel="" type="span" size="large" variation="passive"></oc-icon-stub>
  </button>
  <nav aria-label="Sidebar top navigation menu" class="oc-sidebar-nav oc-my-l oc-px-xs">
    <oc-list-stub>
      <sidebar-nav-item-stub name="All files" index="0" active="true" target="/files/list/all" icon="folder" filltype="fill"></sidebar-nav-item-stub>
      <sidebar-nav-item-stub name="Shared with me" index="1" target="/files/list/shared-with-me" icon="shared-with-me" filltype="fill"></sidebar-nav-item-stub>
      <sidebar-nav-item-stub name="Shared with others" index="2" target="/files/list/shared-with-others" icon="shared-with-others" filltype="fill"></sidebar-nav-item-stub>
      <sidebar-nav-item-stub name="Shared via link" index="3" target="/files/list/shared-via-link" icon="link" filltype="fill"></sidebar-nav-item-stub>
      <sidebar-nav-item-stub name="Deleted files" index="4" target="/files/list/trash-bin" icon="delete" filltype="fill"></sidebar-nav-item-stub>
    </oc-list-stub>
  </nav> <span class="footer">Footer</span>
</div>
`;

exports[`OcSidebarNav toggles into closed state upon button click 1`] = `
<div id="web-nav-sidebar" class="oc-app-navigation-collapsed"><button aria-label="Toggle sidebar" class="toggle-sidebar-button oc-py-s oc-button oc-button-m oc-button-justify-content-center oc-button-gap-m oc-button-inverse oc-button-inverse-raw toggle-sidebar-button-collapsed">
    <oc-icon-stub name="arrow-drop-right" filltype="line" accessiblelabel="" type="span" size="large" variation="passive"></oc-icon-stub>
  </button>
  <nav aria-label="Sidebar top navigation menu" class="oc-sidebar-nav oc-my-l oc-px-xs">
    <oc-list-stub>
      <sidebar-nav-item-stub name="All files" index="0" active="true" target="/files/list/all" icon="folder" filltype="fill" collapsed="true"></sidebar-nav-item-stub>
      <sidebar-nav-item-stub name="Shared with me" index="1" target="/files/list/shared-with-me" icon="shared-with-me" filltype="fill" collapsed="true"></sidebar-nav-item-stub>
      <sidebar-nav-item-stub name="Shared with others" index="2" target="/files/list/shared-with-others" icon="shared-with-others" filltype="fill" collapsed="true"></sidebar-nav-item-stub>
      <sidebar-nav-item-stub name="Shared via link" index="3" target="/files/list/shared-via-link" icon="link" filltype="fill" collapsed="true"></sidebar-nav-item-stub>
      <sidebar-nav-item-stub name="Deleted files" index="4" target="/files/list/trash-bin" icon="delete" filltype="fill" collapsed="true"></sidebar-nav-item-stub>
    </oc-list-stub>
  </nav> <span class="footer">Footer</span>
</div>
`;
